### ISSUE TITLE
Suggestion: entities and models

### ISSUE THREAD

**Issue by RWOverdijk** [balderdashy/waterline#547]
Currently sails (and waterline) rely heavily on the term `model`. This means that the model files contain both definition and logic, which I think is bad practise.

I know this is a lot to apply as changes, but the idea I have is to take the pattern as used in doctrine2 (which architecturally is pretty neat). There you have:

**An entity.**
The entity is simply the definition of, what we now call, the model. It holds the attributes, and optionally getters and setters so the data can be manipulated.

**A repository.**
The repository is used to implement custom `find*` methods. This allows you to make an API for your entity to do stuff that requires multiple lines of code (as to keep the controller thin).

Now, you'd have this (example):

```javascript
// Returns the repository, with the custom find methods.
var user = sails.waterline.getRepository('user');

// Call a custom method as defined in a repository
user.findAssociatedWith('companyName', function(error, entities) {
  // Entities now holds an array of entity objects, composed via the definition supplied.
});
```

The benefits here are:
* That it becomes very clear what's responsible for what. The term `model` is very broad, and could be anything.
* The Entity remains the definition. There's nothing to potentially confuse you in the entity file.
* The repository can be extended easily, and gets its own responsibilities.

To me personally, this would just mean clarity and separation.

**Comment by RWOverdijk** [comment]
@avimar That's true for smaller projects. Larger (data intensive) projects sometimes require a bit more logic. The best place to put custom query logic is in the repository as to make sure you have a place for everything and logic is applied.

**Comment by dudemelo** [comment]
As said @RWOverdijk , the term *model* can be anything. Although not a good practice the same file contain *definition* and *logic*, the Sails *model* is useful for smaller projects, when the business rule are just some *custom selects*. But that does not happen on larger projects involving a complex business rule.

I'm currently working on a project that has a very complex business logic, for this, we distributed our rules in factories, strategies, repositories, chain of responsabilities...

:+1:

**Comment by sailsbot** [comment]
Thanks for posting, @RWOverdijk.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
